File: src/main/java/info/skyblond/archivedag/controller/http/CertController.java
package info.skyblond.archivedag.controller.http;

import info.skyblond.archivedag.model.PermissionDeniedException;
import info.skyblond.archivedag.model.ao.CertChangeStatusRequest;
import info.skyblond.archivedag.model.bo.CertDetailModel;
import info.skyblond.archivedag.model.bo.CertSigningResult;
import info.skyblond.archivedag.service.impl.CertService;
import info.skyblond.archivedag.util.Constants;
import info.skyblond.archivedag.util.GeneralKt;
import info.skyblond.archivedag.util.SecurityUtils;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.sql.Timestamp;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Slf4j
@RestController
@RequestMapping("/cert")
public class CertController {
    private final CertService certService;

    public CertController(CertService certService) {
        this.certService = certService;
    }

    @GetMapping("/signNewCert")
    @PreAuthorize("hasRole('USER')")
    public Map<String, String> signNewCert() {
        String username = SecurityUtils.getCurrentUsername();
        CertSigningResult signingResult;
        try {
            signingResult = this.certService.signCert(username);
        } catch (Throwable t) {
            throw new RuntimeException("Failed to sign new certification", t);
        }

        Map<String, String> result = new HashMap<>();
        result.put("cert", GeneralKt.writeSignCertToString(signingResult));
        result.put("private_key", GeneralKt.writeSignKeyToString(signingResult));
        return result;
    }

    @GetMapping("/listCertSerialNumber")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public List<String> listCertSerialNumber(
            @RequestParam(name = "owner", required = false) String owner,
            @RequestParam(name = "issue_start", required = false) Timestamp issueStart,
            @RequestParam(name = "issue_end", required = false) Timestamp issueEnd,
            @RequestParam(name = "expire_start", required = false) Timestamp expireStart,
            @RequestParam(name = "expire_end", required = false) Timestamp expireEnd,
            Pageable pageable
    ) {
        SecurityUtils.requireSortPropertiesInRange(pageable, List.of("username", "serialNumber", "issuedTime", "expiredTime", "status"));
        if (owner == null) {
            owner = SecurityUtils.getCurrentUsername();
        }
        if (issueStart == null) {
            issueStart = Constants.MIN_TIMESTAMP;
        }
        if (expireStart == null) {
            expireStart = Constants.MIN_TIMESTAMP;
        }
        if (issueEnd == null) {
            issueEnd = Constants.MAX_TIMESTAMP;
        }
        if (expireEnd == null) {
            expireEnd = Constants.MAX_TIMESTAMP;
        }
        // check permission
        if (!SecurityUtils.checkCurrentUserIsAdmin()
                && !SecurityUtils.getCurrentUsername().equals(owner)) {
            // if is not admin, and owner not match
            throw new PermissionDeniedException("You can only list your own certifications");
        }

        if (SecurityUtils.checkCurrentUserIsAdmin()) {
            return this.certService.listCertSerialNumber(true, owner,
                    issueStart, issueEnd, expireStart, expireEnd, pageable);
        } else {
            // if is user
            return this.certService.listCertSerialNumber(false, owner,
                    issueStart, issueEnd, expireStart, expireEnd, pageable);
        }
    }

    @GetMapping("/queryCert")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public CertDetailModel queryCert(
            @RequestParam(name = "serial_number") String serialNumber
    ) {
        // check permission
        if (!SecurityUtils.checkCurrentUserIsAdmin()
                && !this.certService.userOwnCert(SecurityUtils.getCurrentUsername(), serialNumber)) {
            // if is not admin, and not own this cert
            throw new PermissionDeniedException("You can only query your own certifications");
        }
        return this.certService.queryCert(serialNumber);
    }

    @PostMapping("/changeCertStatus")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<?> changeCertStatus(
            @RequestBody CertChangeStatusRequest request
    ) {
        // check permission
        if (!SecurityUtils.checkCurrentUserIsAdmin()
                && !this.certService.userOwnCert(SecurityUtils.getCurrentUsername(),
                request.getSerialNumber())) {
            // if is not admin, and not own this cert
            throw new PermissionDeniedException("You can only change your own certifications");
        }
        this.certService.changeCertStatus(request.getSerialNumber(), request.getNewStatus());
        // return 204
        return ResponseEntity.status(HttpStatus.NO_CONTENT).body(null);
    }

    @DeleteMapping("/deleteCert")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<?> deleteCert(
            @RequestParam("serial_number") String serialNumber
    ) {
        // check permission
        if (!SecurityUtils.checkCurrentUserIsAdmin()
                && !this.certService.userOwnCert(SecurityUtils.getCurrentUsername(),
                serialNumber)) {
            // if is not admin, and not own this cert
            throw new PermissionDeniedException("You can only delete your own certifications");
        }
        this.certService.deleteCert(serialNumber);
        // return 204
        return ResponseEntity.status(HttpStatus.NO_CONTENT).body(null);
    }
}

File: backend/src/main/kotlin/info/skyblond/archivedag/arudaz/controller/http/CertController.kt
package info.skyblond.archivedag.arudaz.controller.http

import info.skyblond.archivedag.arstue.CertService
import info.skyblond.archivedag.arstue.model.CertDetailModel
import info.skyblond.archivedag.arstue.model.CertSigningResult
import info.skyblond.archivedag.arudaz.model.CertChangeStatusRequest
import info.skyblond.archivedag.arudaz.utils.checkCurrentUserIsAdmin
import info.skyblond.archivedag.arudaz.utils.getCurrentUsername
import info.skyblond.archivedag.arudaz.utils.requireSortPropertiesInRange
import info.skyblond.archivedag.commons.PermissionDeniedException
import org.bouncycastle.openssl.jcajce.JcaPEMWriter
import org.bouncycastle.util.io.pem.PemObject
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.io.ByteArrayOutputStream

@RestController
@RequestMapping("/cert")
class CertController(
    private val certService: CertService
) {
    private fun writeSignCertToString(result: CertSigningResult): String {
        val outputStream = ByteArrayOutputStream()
        val pemWriter = JcaPEMWriter(outputStream.writer(Charsets.UTF_8))
        pemWriter.writeObject(result.certificate)
        pemWriter.close()
        return String(outputStream.toByteArray(), Charsets.UTF_8)
    }

    private fun writeSignKeyToString(result: CertSigningResult): String {
        val outputStream = ByteArrayOutputStream()
        val pemWriter = JcaPEMWriter(outputStream.writer(Charsets.UTF_8))
        pemWriter.writeObject(PemObject("PRIVATE KEY", result.privateKey.encoded))
        pemWriter.close()
        return String(outputStream.toByteArray(), Charsets.UTF_8)
    }

    @GetMapping("/signNewCert")
    @PreAuthorize("hasRole('USER')")
    fun signNewCert(): Map<String, String> {
        val username = getCurrentUsername()
        val signingResult: CertSigningResult = try {
            certService.signCert(username)
        } catch (t: Throwable) {
            throw RuntimeException("Failed to sign new certification", t)
        }
        val result: MutableMap<String, String> = HashMap()
        result["serial_number"] = signingResult.serialNumber
        result["cert"] = writeSignCertToString(signingResult)
        result["private_key"] = writeSignKeyToString(signingResult)
        return result
    }

    @GetMapping("/listCertSerialNumber")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    fun listCertSerialNumber(
        @RequestParam(name = "owner", required = false) owner: String?,
        pageable: Pageable?
    ): List<String> {
        requireSortPropertiesInRange(
            pageable!!,
            listOf("username", "serialNumber", "issuedTime", "expiredTime", "status")
        )

        return if (owner.isNullOrBlank()) {
            // owner is empty, return current user's cert
            certService.listCertSerialNumber(
                blurOwner = false,
                owner = getCurrentUsername(),
                pageable = pageable
            )
        } else {
            if (!checkCurrentUserIsAdmin()) {
                // if is not admin, and owner not match
                throw PermissionDeniedException("You can only list your own certifications")
            }
            certService.listCertSerialNumber(
                blurOwner = true,
                owner = owner,
                pageable = pageable
            )
        }
    }

    @GetMapping("/queryCert")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    fun queryCert(
        @RequestParam(name = "serial_number") serialNumber: String
    ): CertDetailModel {
        // check permission
        if (!checkCurrentUserIsAdmin()
            && !certService.userOwnCert(getCurrentUsername(), serialNumber)
        ) {
            // if is not admin, and not own this cert
            throw PermissionDeniedException("You can only query your own certifications")
        }
        return certService.queryCert(serialNumber)
    }

    @PostMapping("/changeCertStatus")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    fun changeCertStatus(
        @RequestBody request: CertChangeStatusRequest
    ): ResponseEntity<*> {
        // check permission
        if (!checkCurrentUserIsAdmin()
            && !certService.userOwnCert(getCurrentUsername(), request.serialNumber)
        ) {
            // if is not admin, and not own this cert
            throw PermissionDeniedException("You can only change your own certifications")
        }
        certService.changeCertStatus(request.serialNumber, request.newStatus)
        // return 204
        return ResponseEntity.status(HttpStatus.NO_CONTENT).body<Any>(null)
    }

    @DeleteMapping("/deleteCert")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    fun deleteCert(
        @RequestParam("serial_number") serialNumber: String
    ): ResponseEntity<*> {
        // check permission
        if (!checkCurrentUserIsAdmin()
            && !certService.userOwnCert(getCurrentUsername(), serialNumber)
        ) {
            // if is not admin, and not own this cert
            throw PermissionDeniedException("You can only delete your own certifications")
        }
        certService.deleteCert(serialNumber)
        // return 204
        return ResponseEntity.status(HttpStatus.NO_CONTENT).body<Any>(null)
    }
}
